### DISCUSSION TITLE
.Net 8 Support

### DISCUSSION THREAD

**Discussion by toddbanister** [QuestPDF/QuestPDF#738] [General]
Does QuestPDF fully support .Net 8 at this time (specifically in displaying images)? 

I have a gaming project that I migrated from .Net 7 to 8 over the weekend and now the images within the PDF are displaying with vastly wrong coloring and such. I still have a copy of my .Net 7 project and reran it and its images still display correctly (examples of the differences below).

<img width="958" alt="Image Examples" src="https://github.com/QuestPDF/QuestPDF/assets/29407045/a70789ec-58a0-498b-b787-09737c5ce9ce">

Both projects download the same images from a Google storage repository and use QuestPDF 2023.12.1 to generate the PDF - but the projects are generating vastly different results.

Additionally - any idea what the squiggly things are in the upper left-hand corner of each bad image? I checked the complete PDF and each image is showing this (and it's obviously not in the correct .Net 7 version shown in the top examples). I assume this is a rendering error of some sort but wasn't sure if it provided any clues as to what might be going on.

Thanks,
TB

**Comment by MarcinZiabek** [comment]
Thank you for raising this issue :) 

QuestPDF does fully support .NET 8. Libraries are not usually required to build specific targets for newer dotnet versions.

This problem is an issue in SkiaSharp 2.88.6 with PNG file format experienced only on arm64 devices. This upgrade was required, as the 2.88.6 version fixes a significant security issue related to JPEG encoding. 

You can ignore this problem if your production environment is a standard x86 machine or downgrade to a previous QuestPDF version for now. I am very sorry for this inconvenience. 

I am surprised that Microsoft is not taking more immediate actions to fix SkiaSharp by updating to newer Skia revisions. I suspect that SkiaSharp is alive thanks to the love of its maintainers rather than funding from Microsoft. I plan to switch to a custom QuestPDF-specific Skia build soon, so this dependency-chain problem should be resolved starting the next release. More information can be found at discussion https://github.com/QuestPDF/QuestPDF/discussions/622

**Reply by toddbanister** [reply]
Thank you very much for the quick reply. 

I am indeed running an ARM64 machine, but the images that are being displayed incorrectly are actually JPGs. And it seems strange that it works correctly on the same machine/code with .Net 7 but not with .Net 8. But regardless, the issue as you described is with SkiaSharp.

If I wanted to try a previous/lower version of QuestPDF (that doesn't rely on  SkiaSharp 2.88.6) until you release a newer version, which one do you suggest?

Thanks for the fantastic product,
TB

**Reply by MarcinZiabek** [reply]
My bad! Indeed, the issue is related to JPEG images, while PNG does work correctly (at least in my testing done on Apple M2). https://github.com/mono/SkiaSharp/issues/2643

I suggest using the [2023.9.1](https://www.nuget.org/packages/QuestPDF/2023.9.1) release.

Btw. I really like your cards design!

**Reply by toddbanister** [reply]
I tried rolling back to 2013.9.1 (which seems to use SkiaSharp 2.88.3) but it was chucking errors talking about the ARM64 platform - so I just shelved my .Net 8 version of the project for now and moved back to the working .Net 7 for the time being. I'll keep an eye out for your next version that might fix the SkiaSharp issue. Thank again for all of the help.

And I am glad you like the design of the cards. :-)

Take care,
TB

**Reply by toddbanister** [reply]
No rush at all - but I was wondering if there was a tentative timeline to getting the SkiaSharp/ARM64 issue resolved. I only ask because .Net 8 has some significant performance updates when using EFCore 8.0. I have noticed my queries are running about 60% faster when using .Net 8 vs. .Net 7.

Again, no rush at all - and as always, I greatly appreciate all of the work with QuestPDF.

TB

**Reply by MarcinZiabek** [reply]
I hope to release a beta version in January, while I anticipate a stable release in February. The idea is to remove the SkiaSharp library altogether and rely on a QuestPDF-specific build.

I am also impressed by the dotnet 8! It improves FluentAPI performance twice, as far as I remember, without any code changes.

**Reply by toddbanister** [reply]
I upgraded to QuestPDF 2023.12.3 today and rebuilt my .Net 8 project. Unfortunately, the image issue with SkiaSharp still remains. I checked the DLL versions and indeed, they are correct (SkiaSharp was 2.88.7). Is there anything I need to do via code to change this behavior? Or was upgrading to the latest QuestPDF all I needed to do?

If if helps, I am using Jebrains Rider on a MacBook Pro (using a M1 Max processor) with .Net 8.0.1.

Thanks for all of the help,
TB

**Reply by MarcinZiabek** [reply]
Would you please ensure the library is updated in your entire solution and/or clean you build?

I have confirmed that SkiaSharp 2.88.7 fixes the JPEG encoding issue on my MacBook M2 Max for both dotnet 6 and 8.

Also, dotnet 8 provide 8% better performance compared to dotnet 6, without any code changes. Quite impressive!

**Reply by toddbanister** [reply]
Long story short, my existing .Net 8 solution simply would not work correctly not matter what I did. I tried damn near everything I knew to get the new QuestPDF to work in that solution. Eventually, I recreated the entire solution from the ground up, using only the 2023.12.3 version, and migrated all of the existing code into the new empty projects. Once I did this, the jpg issue with the ARM processor was finally resolved! Everything seems to be working beautifully at this point with .Net 8.0.1. I now have the QuestPDFs being build correctly and the speed enhancements using EFCore 8.0.

Thank you for all of your help with this issue.

TB

**Reply by MarcinZiabek** [reply]
Have you tried manually removing all `bin` and `obj` directories? It usually helps. Also, sometimes, there are invalid caches in the local nuget storage, so you can remove those too. It happens once a year on my private PC and Mac. But corporate laptops with tons of antivirus programs are the opposite story for me.

Technology... :D

**Comment by DawidRobertNowak** [comment]
Hello,

We are closing this thread due to inactivity. Thank you for taking the time to reach out and contribute to the project — your feedback helps us improve QuestPDF.

Should you encounter any further problems, please don't hesitate to open a new issue.

Thank you!
